pkg/network/master/egressip.go

```diff
@@ -0,0 +1,95 @@
+package master
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	"github.com/openshift/origin/pkg/network/common"
+	networkinformers "github.com/openshift/origin/pkg/network/generated/informers/internalversion/network/internalversion"
+	networkclient "github.com/openshift/origin/pkg/network/generated/internalclientset"
+)
+
+type egressIPManager struct {
+	sync.Mutex
+
+	tracker            *common.EgressIPTracker
+	networkClient      networkclient.Interface
+	hostSubnetInformer networkinformers.HostSubnetInformer
+
+	updatePending bool
+	updatedAgain  bool
+}
+
+func newEgressIPManager() *egressIPManager {
+	eim := &egressIPManager{}
+	eim.tracker = common.NewEgressIPTracker(eim)
+	return eim
+}
+
+func (eim *egressIPManager) Start(networkClient networkclient.Interface, hostSubnetInformer networkinformers.HostSubnetInformer, netNamespaceInformer networkinformers.NetNamespaceInformer) {
+	eim.networkClient = networkClient
+	eim.hostSubnetInformer = hostSubnetInformer
+	eim.tracker.Start(hostSubnetInformer, netNamespaceInformer)
+}
+
+func (eim *egressIPManager) UpdateEgressCIDRs() {
+	eim.Lock()
+	defer eim.Unlock()
+
+	if eim.updatePending {
+		eim.updatedAgain = true
+	} else {
+		eim.updatePending = true
+		go utilwait.PollInfinite(time.Second, eim.maybeDoUpdateEgressCIDRs)
+	}
+}
+
+func (eim *egressIPManager) maybeDoUpdateEgressCIDRs() (bool, error) {
+	eim.Lock()
+	defer eim.Unlock()
+
+	if eim.updatedAgain {
+		eim.updatedAgain = false
+		return false, nil
+	}
+	eim.updatePending = false
+
+	allocation := eim.tracker.ReallocateEgressIPs()
+	for nodeName, egressIPs := range allocation {
+		resultErr := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			hs, err := eim.hostSubnetInformer.Lister().Get(nodeName)
+			if err != nil {
+				return err
+			}
+
+			hs.EgressIPs = egressIPs
+			_, err = eim.networkClient.Network().HostSubnets().Update(hs)
+			return err
+		})
+		if resultErr != nil {
+			utilruntime.HandleError(fmt.Errorf("Could not update HostSubnet EgressIPs: %v", resultErr))
+		}
+	}
+
+	return true, nil
+}
+
+func (eim *egressIPManager) ClaimEgressIP(vnid uint32, egressIP, nodeIP string) {
```

Comment:
why are these functions no-ops? If there's a good reason, could you add a comment?

Reply:
It's part of the `common.EgressIPWatcher` interface. I was trying to share code between the node and master, and ended up doing it in a bad way. I'm going to refactor this later. For now though, the master needs to provide no-op implementations of these methods, and the node needs to provide a no-op implementation of `UpdateEgressIPs`.